### PR TITLE
fix(lexer): do not treat '#bun' in a url as a pragma

### PR DESCRIPTION
**Pull request by DonIsaac**
### What does this PR do?

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?
Added test cases

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 5:44 PM PT - Dec 19th, 2024</sup></div>

Your commit 153d99cdbadc032144c2710eebad678df2155f73 is building: [`#8286`](https://buildkite.com/bun/bun/builds/8286)

File: src/js_lexer.zig [hunk 1]
```diff
@@ -2012,8 +2015,9 @@ fn NewLexer_(
                             }
                         }
 
+                        // @bun must be the first pragma. it may have spaces before it.
                         if (lexer.bun_pragma == .none and strings.hasPrefixWithWordBoundary(chunk, "bun")) {
```
Comment (Jarred-Sumner) [inline src/js_lexer.zig]: This must be a bug in hasPrefixWithWordBoundary. That is supposed to only apply when there is a space or tab character between. it does another pass through the input to check.

**Review by paperclover — CHANGES_REQUESTED** [review]
the fix should just remove the bun_pragma field from the lexer entirely. right now bun is checking every comment for `@bun` when it only needs to check the first one in the file.

**Comment by DonIsaac** [comment]
> the fix should just remove the bun_pragma field from the lexer entirely. right now bun is checking every comment for `@bun` when it only needs to check the first one in the file.

Doing this requires us to have some lexing logic in `js_parser.zig`. Are we ok with this?

**Comment by Jarred-Sumner** [comment]
> Doing this requires us to have some lexing logic in js_parser.zig. Are we ok with this?

We could have a 'has_seen_any_comments' boolean and then not check after that

File: src/js_lexer.zig [hunk 2]
```diff
@@ -2012,8 +2007,9 @@ fn NewLexer_(
...
```
Comment (paperclover) [inline src/js_lexer.zig]: you should be able to remove `bun_pragma`, now that the value is never being read by any code. after removing the field, the changes in this section that write to `bun_pragma` can then be eliminated. the `has_seen_text` part also becomes dead.

**Review by paperclover — CHANGES_REQUESTED** [review]
one nitpick

File: src/js_lexer.zig (hunk 2 above)
Comment (DonIsaac) [reply src/js_lexer.zig]: I can't: `lexer.start != 0`

**Review by paperclover — APPROVED** [review]

**Pull request merged**